snappy/faststreams.nim

```diff
@@ -0,0 +1,147 @@
+import
+  std/strutils,
+  pkg/faststreams/[inputs, multisync, outputs],
+  "."/[codec, encoder, exceptions],
+  ../snappy
+
+export
+  inputs, multisync, outputs, codec, exceptions
+
+{.push raises: [Defect].}
+
+proc checkCrcAndAppend(
+    output: OutputStream, data: openArray[byte], crc: uint32): bool {.
+    raises: [Defect, IOError].}=
+  if maskedCrc(data) == crc:
+    output.write(data)
+    return true
+
+proc compress*(input: InputStream, output: OutputStream) {.
+    raises: [Defect, InputTooLarge, IOError].} =
+  ## Compress all bytes of `input`, writing into `output` and flushing at the end.
+  ##
+  ## Input length must not exceed `maxUncompressedLen == 2^32-1` or
+  ## `InputTooLarge` will be raised. Other errors are raised as they happen on
+  ## the given streams.
+  doAssert input.len.isSome, "TODO: support actual .. streams"
+  let
+    lenU32 = checkInputLen(input.len.get).valueOr:
+      raiseInputTooLarge()
+    maxCompressed = maxCompressedLen(input.len.get).valueOr:
+      raiseInputTooLarge()
+
+  # TODO https://github.com/status-im/nim-faststreams/issues/28
+  # output.ensureRunway maxCompressed
+  output.write lenU32.toBytes(Leb128).toOpenArray()
+
+  var
+    # TODO instead of a temporary buffer, use `getWriteableBytes` once it
+    #      works
+    tmp = newSeqUninitialized[byte](int maxCompressedBlockLen)
+
+  while input.readable(maxBlockLen.int):
+    let written = encodeBlock(input.read(maxBlockLen.int), tmp)
+    # TODO async streams could be supported efficiently by waiting here, after
```

Comment:
You can use the `fsMultiSync` macro to compile this code for both sync and async streams.

Reply:
perhaps, but there are not tests for this in the existing code, so it's outside of the scope of the PR

Reply:
also .. I'm not sure actually what else needs to be done / called - I've just seen that there were some experiments in the code using `fsMultiSync` - the synchronization point was places at the worst place possible though, in the middle of a hotspot of a tight loop that outputs only a few bytes per iteration - synchronizing there would have been quite bad.